File: examples/python/AutoCalibration/auto_calibration_example.py
import cv2 as cv
import numpy as np
import depthai as dai
import warnings


def showDepth(depthFrame, windowName="Depth", minDistance=500, maxDistance=5000,
               colormap=cv.COLORMAP_TURBO, useLog=False):
    """
    Nicely visualize a depth map.

    Args:
        depthFrame (np.ndarray): Depth frame (in millimeters).
        windowName (str): OpenCV window name.
        minDistance (int): Minimum depth to display (in mm).
        maxDistance (int): Maximum depth to display (in mm).
        colormap (int): OpenCV colormap (e.g., cv.COLORMAP_JET, COLORMAP_TURBO, etc.).
        useLog (bool): Apply logarithmic scaling for better visual contrast.
    """
    if maxDistance <= minDistance:
        warnings.warn(
            f"Invalid distance range: maxDistance ({maxDistance}) <= minDistance ({minDistance})",
            stacklevel=2,
        )
        return

    # Convert to float for processing
    depthFrame = depthFrame.astype(np.float32)

    # Optionally apply log scaling
    if useLog:
        depthFrame = np.log(depthFrame + 1)
        minDistance = np.log(minDistance + 1)
        maxDistance = np.log(maxDistance + 1)

    # Clip and normalize to [0, 255]
    depthFrame = np.clip(depthFrame, minDistance, maxDistance)
    depthFrame = np.uint8((depthFrame - minDistance) * (255.0 / (maxDistance - minDistance)))

    # Apply color map
    depthColor = cv.applyColorMap(depthFrame, colormap)

    # Show in a window
    cv.imshow(windowName, depthColor)


def rotationMatrixToEulerAngles(rotationMatrix, vector=False):
    if vector:
        try:
            import cv2 as _cv2
        except ImportError as exc:
            raise RuntimeError("vector=True requires opencv-python (cv2)") from exc
        rotationMatrix, _ = _cv2.Rodrigues(rotationMatrix)

    sy = np.sqrt(rotationMatrix[0, 0] ** 2 + rotationMatrix[1, 0] ** 2)
    singular = sy < 1e-6
    if not singular:
        xAngle = np.arctan2(rotationMatrix[2, 1], rotationMatrix[2, 2])
        yAngle = np.arctan2(-rotationMatrix[2, 0], sy)
        zAngle = np.arctan2(rotationMatrix[1, 0], rotationMatrix[0, 0])
    else:
        xAngle = np.arctan2(-rotationMatrix[1, 2], rotationMatrix[1, 1])
        yAngle = np.arctan2(-rotationMatrix[2, 0], sy)
        zAngle = 0.0

    return np.rad2deg(xAngle), np.rad2deg(yAngle), np.rad2deg(zAngle)


def eulerAnglesToRotationMatrix(phi, theta, psi):
    phi = np.radians(phi)
    theta = np.radians(theta)
    psi = np.radians(psi)

    rx = np.array(
        [
            [1, 0, 0],
            [0, np.cos(phi), -np.sin(phi)],
            [0, np.sin(phi), np.cos(phi)],
        ]
    )
    ry = np.array(
        [
            [np.cos(theta), 0, np.sin(theta)],
            [0, 1, 0],
            [-np.sin(theta), 0, np.cos(theta)],
        ]
    )
    rz = np.array(
        [
            [np.cos(psi), -np.sin(psi), 0],
            [np.sin(psi), np.cos(psi), 0],
            [0, 0, 1],
        ]
    )
    return np.dot(rz, np.dot(ry, rx))


def botchCalibration(device : dai.Device):
    calibrationHandler = device.readCalibration()
    extrinsics = calibrationHandler.getCameraExtrinsics(
        dai.CameraBoardSocket.CAM_B, dai.CameraBoardSocket.CAM_C
    )

    extrinsicsNp = np.array(extrinsics, dtype=float)
    rCurrent = extrinsicsNp[0:3, 0:3]
    tCurrent = extrinsicsNp[0:3, 3]

    phi, theta, psi = rotationMatrixToEulerAngles(rCurrent)
    phiBotched = phi + 0.15
    thetaBotched = theta + 0.10
    psiBotched = psi
    rBotched = eulerAnglesToRotationMatrix(phiBotched, thetaBotched, psiBotched)

    t = [float(tCurrent[0]), float(tCurrent[1]), float(tCurrent[2])]
    r = rBotched.tolist()

    print(f"Original Euler deg (x, y, z): ({phi:.6f}, {theta:.6f}, {psi:.6f})")
    print(f"Botched  Euler deg (x, y, z): ({phiBotched:.6f}, {thetaBotched:.6f}, {psiBotched:.6f})")
    print(f"Keeping translation vector (cm): {t}")

    calibrationHandler.setCameraExtrinsics(
        dai.CameraBoardSocket.CAM_B, dai.CameraBoardSocket.CAM_C, r, t, t
    )
    device.setCalibration(calibrationHandler)


# Create pipeline
with dai.Pipeline() as pipeline:
    device = pipeline.getDefaultDevice()
    botchCalibration(device)

    camLeft = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_B)
    camRight = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_C)
    stereo = pipeline.create(dai.node.StereoDepth)

    dcWorker = pipeline.create(dai.node.AutoCalibration).build(camLeft, camRight)
    dcWorker.initialConfig.maxIterations = 2
    dcWorker.initialConfig.sleepingTime = 10
    dcWorker.initialConfig.flashCalibration = False
    dcWorker.initialConfig.mode = dai.AutoCalibrationConfig.CONTINUOUS  # ON_START
    dcWorker.initialConfig.validationSetSize = 5
    dcWorker.initialConfig.dataConfidenceThreshold = 0.3
    workerOutputQueue = dcWorker.output.createOutputQueue()

    videoQueueLeft = camLeft.requestOutput((1280, 800), fps=30)
    videoQueueRight = camRight.requestOutput((1280, 800), fps=30)

    videoQueueLeft.link(stereo.left)
    videoQueueRight.link(stereo.right)

    stereoOut = stereo.depth.createOutputQueue()
    pipeline.start()

    while pipeline.isRunning():
        workerOutput = workerOutputQueue.tryGet()
        if workerOutput is not None:
            if workerOutput.passed:
                print("Passed")
                print(f"dataConfidence = {workerOutput.dataConfidence}")
                print(f"calibrationConfidence = {workerOutput.calibrationConfidence}")
            else:
                print("Did not pass")

        depth = stereoOut.get()
        showDepth(
            depth.getCvFrame(),
            windowName="Depth",
            minDistance=500,
            maxDistance=5000,
            colormap=cv.COLORMAP_TURBO,
            useLog=False
        )

        if cv.waitKey(1) == ord("q"):
            break

    pipeline.stop()

File: examples/cpp/AutoCalibration/auto_calibration_example.cpp
#include <cmath>
#include <iomanip>
#include <iostream>
#include <memory>
#include <opencv2/opencv.hpp>
#include <stdexcept>
#include <tuple>
#include <vector>

#include "depthai/depthai.hpp"

// Visualization helper
void showDepth(const cv::Mat& depthFrame, const std::string& windowName = "Depth", int minDistance = 500, int maxDistance = 5000) {
    if(maxDistance <= minDistance) return;

    cv::Mat clipped = depthFrame.clone();
    clipped.setTo(minDistance, depthFrame < minDistance);
    clipped.setTo(maxDistance, depthFrame > maxDistance);

    cv::Mat displayFrame;
    double scale = 255.0 / (maxDistance - minDistance);
    double offset = -minDistance * scale;
    clipped.convertTo(displayFrame, CV_8UC1, scale, offset);

    cv::Mat colorMap;
    cv::applyColorMap(displayFrame, colorMap, cv::COLORMAP_TURBO);

    cv::imshow(windowName, colorMap);
}

std::tuple<double, double, double> rotationMatrixToEulerAngles(const cv::Matx33d& rotationMatrix) {
    constexpr double kPi = 3.14159265358979323846;
    const double sy = std::sqrt(rotationMatrix(0, 0) * rotationMatrix(0, 0) + rotationMatrix(1, 0) * rotationMatrix(1, 0));
    const bool singular = sy < 1e-6;

    double xAngle;
    double yAngle;
    double zAngle;
    if(!singular) {
        xAngle = std::atan2(rotationMatrix(2, 1), rotationMatrix(2, 2));
        yAngle = std::atan2(-rotationMatrix(2, 0), sy);
        zAngle = std::atan2(rotationMatrix(1, 0), rotationMatrix(0, 0));
    } else {
        xAngle = std::atan2(-rotationMatrix(1, 2), rotationMatrix(1, 1));
        yAngle = std::atan2(-rotationMatrix(2, 0), sy);
        zAngle = 0.0;
    }

    return {xAngle * 180.0 / kPi, yAngle * 180.0 / kPi, zAngle * 180.0 / kPi};
}

cv::Matx33d eulerAnglesToRotationMatrix(double phi, double theta, double psi) {
    constexpr double kPi = 3.14159265358979323846;
    const double phiRad = phi * kPi / 180.0;
    const double thetaRad = theta * kPi / 180.0;
    const double psiRad = psi * kPi / 180.0;

    const cv::Matx33d rx(1.0, 0.0, 0.0, 0.0, std::cos(phiRad), -std::sin(phiRad), 0.0, std::sin(phiRad), std::cos(phiRad));
    const cv::Matx33d ry(std::cos(thetaRad), 0.0, std::sin(thetaRad), 0.0, 1.0, 0.0, -std::sin(thetaRad), 0.0, std::cos(thetaRad));
    const cv::Matx33d rz(std::cos(psiRad), -std::sin(psiRad), 0.0, std::sin(psiRad), std::cos(psiRad), 0.0, 0.0, 0.0, 1.0);

    return rz * ry * rx;
}

void botchCalibration(std::shared_ptr<dai::Device> device) {
    auto calibHandler = device->readCalibration();
    const auto extrinsics = calibHandler.getCameraExtrinsics(dai::CameraBoardSocket::CAM_B, dai::CameraBoardSocket::CAM_C);

    if(extrinsics.size() < 3 || extrinsics[0].size() < 4 || extrinsics[1].size() < 4 || extrinsics[2].size() < 4) {
        throw std::runtime_error("Invalid CAM_B->CAM_C extrinsics shape, expected at least 3x4.");
    }

    const cv::Matx33d rCurrent(static_cast<double>(extrinsics[0][0]),
                               static_cast<double>(extrinsics[0][1]),
                               static_cast<double>(extrinsics[0][2]),
                               static_cast<double>(extrinsics[1][0]),
                               static_cast<double>(extrinsics[1][1]),
                               static_cast<double>(extrinsics[1][2]),
                               static_cast<double>(extrinsics[2][0]),
                               static_cast<double>(extrinsics[2][1]),
                               static_cast<double>(extrinsics[2][2]));

    auto [phi, theta, psi] = rotationMatrixToEulerAngles(rCurrent);
    const double phiBotched = phi + 0.15;
    const double thetaBotched = theta + 0.10;
    const double psiBotched = psi;

    const cv::Matx33d rBotched = eulerAnglesToRotationMatrix(phiBotched, thetaBotched, psiBotched);

    std::vector<float> t = {extrinsics[0][3], extrinsics[1][3], extrinsics[2][3]};
    std::vector<std::vector<float>> r = {{static_cast<float>(rBotched(0, 0)), static_cast<float>(rBotched(0, 1)), static_cast<float>(rBotched(0, 2))},
                                         {static_cast<float>(rBotched(1, 0)), static_cast<float>(rBotched(1, 1)), static_cast<float>(rBotched(1, 2))},
                                         {static_cast<float>(rBotched(2, 0)), static_cast<float>(rBotched(2, 1)), static_cast<float>(rBotched(2, 2))}};

    std::cout << std::fixed << std::setprecision(6);
    std::cout << "Original Euler deg (x, y, z): (" << phi << ", " << theta << ", " << psi << ")" << std::endl;
    std::cout << "Botched  Euler deg (x, y, z): (" << phiBotched << ", " << thetaBotched << ", " << psiBotched << ")" << std::endl;
    std::cout << "Keeping translation vector (cm): [" << t[0] << ", " << t[1] << ", " << t[2] << "]" << std::endl;

    calibHandler.setCameraExtrinsics(dai::CameraBoardSocket::CAM_B, dai::CameraBoardSocket::CAM_C, r, t, t);
    device->setCalibration(calibHandler);
}

int main() {
    dai::Pipeline pipeline;

    // Create device
    auto device = pipeline.getDefaultDevice();
    botchCalibration(device);

    // Nodes
    auto camLeft = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_B);
    auto camRight = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_C);
    auto stereo = pipeline.create<dai::node::StereoDepth>();

    // AutoCalibration node
    auto dcWorker = pipeline.create<dai::node::AutoCalibration>();
    dcWorker->build(camLeft, camRight);

    auto config = dcWorker->initialConfig;
    config->maxIterations = 2;
    config->sleepingTime = 10;
    config->flashCalibration = false;
    config->mode = dai::AutoCalibrationConfig::CONTINUOUS;
    config->validationSetSize = 5;
    config->dataConfidenceThreshold = 0.3;

    // Links
    camLeft->requestOutput({1280, 800})->link(stereo->left);
    camRight->requestOutput({1280, 800})->link(stereo->right);

    // Queues
    auto workerOutputQueue = dcWorker->output.createOutputQueue();
    auto stereoOut = stereo->depth.createOutputQueue();

    pipeline.start();

    while(pipeline.isRunning()) {
        auto workerOutput = workerOutputQueue->tryGet<dai::AutoCalibrationResult>();
        if(workerOutput != nullptr) {
            if(workerOutput->passed) {
                std::cout << "Passed" << std::endl;
                std::cout << "dataConfidence = " << workerOutput->dataConfidence << std::endl;
                std::cout << "calibrationConfidence = " << workerOutput->calibrationConfidence << std::endl;
            } else {
                std::cout << "Did not pass." << std::endl;
            }
        }

        auto depth = stereoOut->get<dai::ImgFrame>();
        showDepth(depth->getCvFrame(), "Depth", 500, 5000);

        if(cv::waitKey(1) == 'q') break;
    }

    return 0;
}
